src/Components/test/testassets/BasicTestApp/FormsTest/NotifyPropertyChangedValidationComponent.cshtml

```diff
@@ -0,0 +1,105 @@
+@using System.ComponentModel
+@using System.ComponentModel.DataAnnotations
+@using System.Runtime.CompilerServices;
+@using Microsoft.AspNetCore.Components.Forms
+
+<p>
+    There's no requirement for models to implement INotifyPropertyChanged, but if they do,
+    you can easily wire that up to the EditContext. Then you have no need to use the built-in
+    Input* components - you can instead bind to regular HTML elements and still get modification
+    notifications. This provides more flexibility in how the UI is rendered, at the cost of
+    more complexity and boilerplate in your model classes.
+</p>
+<p>
+    This example also shows that you don't strictly have to use EditForm. You can manually
+    cascade an EditContext to the components that integrate with it. 
+</p>
+
+<form onsubmit="@HandleSubmit">
+    <p>
+        User name:
+        <input bind="@person.UserName" class="@editContext.FieldClass(() => person.UserName)" />
+    </p>
+    <p>
+        Accept terms:
+        <input type="checkbox" bind="@person.AcceptsTerms" class="@editContext.FieldClass(() => person.AcceptsTerms)" />
```

Comment:
This is the kind of place where having an extensible directive-attribute kind of system would be useful. You could imagine someone writing their own little helper to pull out the class and using that attribute instead.

Reply:
OK, it's good to include this scenario in the considerations then.